extensions/mssql/src/objectExplorerNodeProvider/objectExplorerNodeProvider.ts

```diff
@@ -182,108 +160,125 @@ export class MssqlObjectExplorerNodeProvider extends ProviderBase implements sql
-				rootNode = node;
+				sqlClusterSession = node.session;
 				break;
 			} else {
 				node = node.parent;
 			}
 		}
-		if (rootNode) {
-			return rootNode.session;
-		}
-		// Not found
-		return undefined;
+		return sqlClusterSession;
 	}
 
-	async findNodeForContext<T extends TreeNode>(explorerContext: sqlops.ObjectExplorerContext): Promise<T> {
+	async findSqlClusterNodeByContext<T extends TreeNode>(context: ICommandObjectExplorerContext | sqlops.ObjectExplorerContext): Promise<T> {
 		let node: T = undefined;
-		let session = await this.findSessionForConnection(explorerContext.connectionProfile);
+		let explorerContext = 'explorerContext' in context ? context.explorerContext : context;
+		let sqlConnProfile = explorerContext.connectionProfile;
+		let session = this.findSqlClusterSessionBySqlConnProfile(sqlConnProfile);
 		if (session) {
 			if (explorerContext.isConnectionNode) {
 				// Note: ideally fix so we verify T matches RootNode and go from there
-				node = <T><any>session.root;
+				node = <T><any>session.rootNode;
 			} else {
 				// Find the node under the session
-				node = <T><any>await session.root.findNodeByPath(explorerContext.nodeInfo.nodePath, true);
+				node = <T><any>await session.rootNode.findNodeByPath(explorerContext.nodeInfo.nodePath, true);
 			}
 		}
 		return node;
 	}
 
-	private async findSessionForConnection(connectionProfile: sqlops.IConnectionProfile): Promise<Session> {
+	public findSqlClusterSessionBySqlConnProfile(connectionProfile: sqlops.IConnectionProfile): SqlClusterSession {
 		for (let session of this.sessionMap.values()) {
-			if (session.connection && await session.connection.isMatch(connectionProfile)) {
+			if (session.isMatchedSqlConnection(connectionProfile)) {
 				return session;
 			}
 		}
 		return undefined;
 	}
 }
 
-export class Session {
-	private _root: RootNode;
-	constructor(private _connection: Connection, private sessionId?: string) {
-	}
+export class SqlClusterSession {
+	private _sqlClusterConnection: SqlClusterConnection;
+	private _sqlSession: sqlops.ObjectExplorerSession;
+	private _sqlConnProfile: sqlops.IConnectionProfile;
+	private _rootNode: SqlClusterRootNode;
 
-	public get uri(): string {
-		return this.sessionId || this._connection.uri;
+	constructor(arg: SqlClusterSessionArgs) {
+		this._sqlClusterConnection = arg.sqlClusterConnection;
+		this._sqlSession = arg.sqlSession;
+		this._sqlConnProfile = arg.sqlConnProfile;
+		this._rootNode = new SqlClusterRootNode(this,
+			new TreeDataContext(arg.appContext.extensionContext, arg.changeHandler),
+			arg.sqlSession.rootNode.nodePath);
 	}
 
-	public get connection(): Connection {
-		return this._connection;
-	}
+	public get sqlClusterConnection(): SqlClusterConnection { return this._sqlClusterConnection; }
+	public get sqlSession(): sqlops.ObjectExplorerSession { return this._sqlSession; }
+	public get sqlConnProfile(): sqlops.IConnectionProfile { return this._sqlConnProfile; }
```

Comment:
Naming: as requested this should be `sqlConnectionProfile` to avoid contractions

Reply:
fixed.